### DISCUSSION TITLE
function call as stimulus

### DISCUSSION THREAD

**Discussion by idoba2030** [jspsych/jsPsych#1313] [Q&A] [answered]
Hey everyone,
Thanks for developing jspsych, I really enjoy working with it.

I need your help in the following matter.
I used the stimulus parameter as a function call that randomly picks images and locations to show on the memory array of the trial. I want to use the output of this function as I need it to then create the target array of the trial which draws upon those images and locations. I wasn't able to get access to the values generated by the random assignment to the stimulus parameter, as every time I call it generates new values.

        var memory_array = function (condition) {
            number_of_images = parseInt(condition[0])
            locations = get_locations(location_names, number_of_images)
            squares_array = get_squares(change_detection_images, number_of_images)
            var squares = '';
            for (index = 0; index < number_of_images; index++) {
                squares += locations[index] + squares_array[index] + ">"
            }
            var target = locations[0] + squares_array[0] + ">"
            return [squares, target]
        }

      var memory = {
            type: "html-keyboard-response",
            stimulus: function () { return memory_array(condition)[0] },
            choices: jsPsych.NO_KEYS,
            trial_duration: 200,
        }

        var test = {
            type: "html-keyboard-response",
            stimulus:/* a function call based on the values of memory.stimulus()*/,
            choices: [83, 75],
            trial_duration: 6000
        }

Thanks for the help!

**Comment by becky-gilbert** [comment, marked as answer]
Hi @idoba2030, one way to do this is to use a global variable to store the current value of your memory array each time it changes in the `memory` trial, and then in your `test` trial, access that variable in the stimulus function, like this:
```javascript
var current_memory_stim = null; // set up a global variable to hold your memory stimulus

var memory_array = function (condition) {
  number_of_images = parseInt(condition[0])
  locations = get_locations(location_names, number_of_images)
  squares_array = get_squares(change_detection_images, number_of_images)
  var squares = '';
  for (index = 0; index < number_of_images; index++) {
    squares += locations[index] + squares_array[index] + ">"
  }
  var target = locations[0] + squares_array[0] + ">"
  // update the value of the global variable each time the memory_array function is called
  current_memory_stim = [squares, target]; 
  return current_memory_stim;
}

var memory = {
  type: "html-keyboard-response",
  stimulus: function () { return memory_array(condition)[0] },
  choices: jsPsych.NO_KEYS,
  trial_duration: 200,
}

var test = {
  type: "html-keyboard-response",
  stimulus: function () {
    // use current_memory_stim to access the memory array from the last trial
    // and use that to generate your stimulus here 
  },
  choices: [83, 75],
  trial_duration: 6000
}
```
I hope this helps. Feel free to post here if you get stuck!

**Reply by idoba2030** [reply]
Hey Becky,
It helps a lot.
I wasn't aware of the fact that I can update the values of the global function from within the function.
I thought that the return command is the only communication it has with the environment.

In the end, I found a different solution here at the forum that I hope will work as well.

        var change_detection_images = ['images/squares/black.png', 'images/squares/blue.png', 'images/squares/brown.png', 'images/squares/cyan.png', 'images/squares/green.png', 'images/squares/magenta.png', 'images/squares/orange.png', 'images/squares/red.png', 'images/squares/yellow.png']
        var location_names = ["<img class='a' src=", "<img class='b' src=", "<img class='c' src=", "<img class='d' src="]
        var get_squares = function (images, number_of_images) {
            return jsPsych.randomization.sampleWithoutReplacement(change_detection_images, number_of_images)
        }
        var get_locations = function (location_names, number_of_images) {
            return jsPsych.randomization.sampleWithoutReplacement(location_names, number_of_images)
        }

        var memory = {
            type: "html-keyboard-response",
            stimulus: function () {
                number_of_images = parseInt(condition[0][0])
                change_detection_images = jsPsych.randomization.shuffle(change_detection_images)
                location_names = jsPsych.randomization.shuffle(location_names)
                var squares = '';
                for (index = 0; index < number_of_images; index++) {
                    squares += location_names[index] + change_detection_images[index] + ">"
                }
                console.log(squares);
                return squares;
            },
            choices: jsPsych.NO_KEYS,
            trial_duration: 3000,
        }


        var test = {
            type: "html-keyboard-response",
            stimulus: function () {
                if (condition[0][1] == 's') { /* it is the same condition */
                    square = location_names[0] + change_detection_images[0] + ">"
                }
                else { /* it is different condition */
                    square = location_names[0] + jsPsych.randomization.sampleWithoutReplacement(change_detection_images, 1) + ">"
                }
                console.log(condition)
                console.log(square)
                return square
            },
            choices: [83, 75],
            trial_duration: 6000,
            on_finish: function () {
                condition = jsPsych.randomization.sampleWithoutReplacement(['4same', '4diff'], 1)
                console.log(condition)
                console.log('-----------')
            }

I seem to be getting what I want... I hope it's ok and I can continue to the next challenge with this code :).
